pkg/download/consistent_hashing.go

```diff
@@ -100,46 +102,59 @@ func (m *ConsistentHashingMode) Fetch(ctx context.Context, urlString string) (io
 		return m.FallbackStrategy.Fetch(ctx, urlString)
 	}
 
-	firstChunkResp, err := m.DoRequest(ctx, 0, m.minChunkSize()-1, urlString)
-	if err != nil {
+	br := newBufferedReader(m.minChunkSize())
+	firstReqResultCh := make(chan firstReqResult)
+	m.queue.submit(func() {
+		m.sem.Go(func() error {
+			defer close(firstReqResultCh)
+			defer br.done()
+			firstChunkResp, err := m.DoRequest(ctx, 0, m.minChunkSize()-1, urlString)
+			if err != nil {
+				firstReqResultCh <- firstReqResult{err: err}
+				return err
+			}
+			defer firstChunkResp.Body.Close()
+
+			trueURL := firstChunkResp.Request.URL.String()
+			if trueURL != urlString {
+				logger.Info().Str("url", urlString).Str("redirect_url", trueURL).Msg("Redirect")
+			}
+
+			fileSize, err := m.getFileSizeFromContentRange(firstChunkResp.Header.Get("Content-Range"))
+			if err != nil {
+				firstReqResultCh <- firstReqResult{err: err}
+				return err
+			}
+			firstReqResultCh <- firstReqResult{fileSize: fileSize, trueURL: trueURL}
+
+			return br.downloadBody(firstChunkResp)
+		})
+	})
+	firstReqResult, ok := <-firstReqResultCh
+	if !ok {
+		panic("shouldn't happen")
```

Comment:
This is a good use of panic but I'd like the message to be descriptive in case it's hit in the real world: "shouldn't happen" is going to require searching the code to find where it came from (Even though I know today).

Reply:
agreed. also note to self: I did panic here but return error in buffer.go, it should be consistent.